### PR TITLE
Add research permissions to captain games

### PR DESCRIPTION
**Pull request by Darkenkade**
### Summary
- Captains can lock research for their team, preventing players from starting, cancelling, or reordering research in the tech tree
- Captains can whitelist trusted players who retain research access while the lock is active
- Uses a `no_research` permission group (clone of Default minus `start_research`, `cancel_research`, `move_research` actions)
- Cleans up permission groups on captain event end or captain removal

### Details
- New UI section in captain manager GUI with lock toggle, trustlist add/remove dropdowns, and status labels
- Follows the same UX pattern as the existing science throwing permissions
- Handles edge cases: player joining a locked team, spectate/rejoin, freeze/unfreeze during picking phase, captain mode cancellation

### Test plan
- [x] Start captain game, pick teams, start playing
- [x] Captain clicks "Lock research" — team players can no longer interact with tech tree
- [x] Captain adds a player to research trustlist — that player can use tech tree again
- [x] Captain removes from trustlist — player blocked again
- [x] Captain clicks "Unlock research" — everyone can use tech tree, trustlist cleared
- [x] Cancel captain event — all players back in Default group, `no_research` group deleted
- [x] Player goes spectate and rejoins locked team — still restricted
- [x] New player picked into locked team — starts restricted

### Tested Changes:
- [x] I've tested the changes with people.

### CODE REVIEW

File: comfy_panel/special_games/captain.lua [hunk 1]
```diff
@@ -128,6 +128,140 @@ local function add_to_trust(playerName)
     end
 end
 
+-- == RESEARCH PERMISSIONS ====================================================
+-- Uses a 'no_research' permission group (clone of Default minus research
+-- actions) to block tech-tree interactions for non-trusted players.
+-- Note: if the blueprint toggle in config.lua fires while research is locked,
+-- the no_research group may desync from Default for blueprint permissions.
+-- This is intentionally not handled — extremely unlikely during captain games.
```
Comment (DaTa-) [inline comfy_panel/special_games/captain.lua]: I suggest you to not neglect this issue and fix it anyway. See next suggestions for more information

File: comfy_panel/special_games/captain.lua [hunk 2]
```diff
@@ -203,6 +340,7 @@ local function force_end_captain_event(player)
         game.print('Captain event was canceled')
     end
 
+    cleanup_research_permissions()
```
Comment (DaTa-) [inline comfy_panel/special_games/captain.lua]: This only cleans up when the special forcefully closed. You also need to do it when the special ends in a more natural way (the match has ended, map has been reset). I think a good place to do it is `function Public.reset_special_games()` at `captain.lua`

File: comfy_panel/special_games/captain.lua [hunk 3]
```diff
@@ -1890,6 +2039,98 @@ local function on_gui_click(event)
...
+                'Research locked for your team: ' .. tostring(special.southResearchLocked),
+                { color = Color.yellow }
+            )
+        end
+        Public.update_all_captain_player_guis()
+    elseif name == 'captain_add_research_trustlist' then
+        local frame = Public.get_active_tournament_frame(player, 'captain_manager_gui')
+        local playerNameUpdateText =
+            get_dropdown_value(frame.captain_manager_research_trustlist_table.captain_add_research_trustlist_playerlist)
+        if playerNameUpdateText and playerNameUpdateText ~= '' and playerNameUpdateText ~= 'Select Player' then
```
Comment (DaTa-) [inline comfy_panel/special_games/captain.lua]: I would avoid using weird `get_dropdown_value`
```suggestion
        local playerlist = frame.captain_manager_research_trustlist_table.captain_add_research_trustlist_playerlist
        if playerlist.selected_index ~= 1 then
            local playerNameUpdateText = playerlist.items[playerlist.selected_index]
```

File: comfy_panel/special_games/captain.lua [hunk 4]
```diff
@@ -128,6 +128,140 @@ local function add_to_trust(playerName)
...
+        group.destroy()
+    end
+end
+
+-- Delayed callback to re-restrict a player after join_team puts them in Default.
+-- on_player_changed_force fires before join_team finishes setting the permission
+-- group, so we schedule a 1-tick delay to apply the restriction after.
+local apply_research_restriction_token = Token.register(function(data)
+    local player = game.get_player(data.player_index)
+    if not player or not player.valid then
```
Comment (DaTa-) [inline comfy_panel/special_games/captain.lua]: Factorio API never returns invalid objects, with the exception of stalled references inside `storage`

File: comfy_panel/special_games/captain.lua [hunk 5]
```diff
@@ -1890,6 +2039,98 @@ local function on_gui_click(event)
...
+                unlock_team_research('south')
+            else
+                special.southResearchLocked = true
+                lock_team_research('south')
+            end
+            game.forces.south.print(
+                'Research locked for your team: ' .. tostring(special.southResearchLocked),
+                { color = Color.yellow }
+            )
+        end
```
Comment (DaTa-) [inline comfy_panel/special_games/captain.lua]: Maybe de-duplicate this code?

File: comfy_panel/special_games/captain.lua [hunk 6]
```diff
@@ -1890,6 +2039,98 @@ local function on_gui_click(event)
...
+                local playerToRemove = cpt_get_player(playerNameUpdateText)
+                if is_research_locked(forceForPrint) and playerToRemove and playerToRemove.valid then
+                    get_or_create_no_research_group().add_player(playerToRemove)
+                end
+                game.forces[forceForPrint].print(
+                    playerNameUpdateText .. ' removed from research trustlist',
+                    { color = Color.green }
+                )
+            else
+                player.print(playerNameUpdateText .. ' is not on the research trustlist', { color = Color.red })
```
Comment (DaTa-) [inline comfy_panel/special_games/captain.lua]: I think `table_contains(trustlist, playerNameUpdateText)` always true and this code is unreachable, unless I'm missing something

**Comment by DaTa-** [comment]
I also found another issue: if you lock research for you team, add a player to the trustlist, switch the player to another team via team manager they will be still in your trustlist, so you can "untrust" them so they will be stuck in 'no_research' group unable to change other teams research queue.

Tip: `on_player_changed_force` is a bad place to react to switching of the team and probably `Public.update_all_captain_player_guis()` run there is broken too (`storage.chosen_team` not yet set properly)

File: comfy_panel/special_games/captain.lua [hunk 7]
```diff
@@ -131,36 +131,38 @@ end
 -- == RESEARCH PERMISSIONS ====================================================
 -- Uses a 'no_research' permission group (clone of Default minus research
 -- actions) to block tech-tree interactions for non-trusted players.
--- Note: if the blueprint toggle in config.lua fires while research is locked,
--- the no_research group may desync from Default for blueprint permissions.
--- This is intentionally not handled — extremely unlikely during captain games.
```
Comment (DaTa-) [inline comfy_panel/special_games/captain.lua]: If you don't want to address blueprint toggle issue, at least mention it somewhere. For now toggling blueprints in Config tab desyncs permissions between 'Default' and 'no_research' group

File: comfy_panel/special_games/captain.lua (hunk 2 above)
Comment (DaTa-) [reply comfy_panel/special_games/captain.lua]: This issue not addressed by the last changes

File: comfy_panel/special_games/captain.lua [hunk 8]
```diff
@@ -939,6 +1081,7 @@ local function start_captain_event()
     if storage.freeze_players == true then
         storage.freeze_players = false
         TeamManager.unfreeze_players()
+        resync_no_research_group()
```
[the remark this answers is not in the document]
Comment (DaTa-) [reply comfy_panel/special_games/captain.lua]: This issue not addressed by the last changes

File: comfy_panel/special_games/captain.lua [hunk 9]
```diff
@@ -2926,6 +2877,10 @@ function Public.draw_captain_manager_gui(player, main_frame)
         type = 'label',
         caption = '[font=heading-1][color=purple]Management for research[/color][/font]',
     })
+    main_frame.add({
+        type = 'label',
+        caption = 'Lock research for your team. Trusted players can still change research. Captains and vice captains can manage this.',
+    })
```
Comment (DaTa-) [inline comfy_panel/special_games/captain.lua]: This made my captain's team manager ui ultra wide:

Image

File: comfy_panel/special_games/captain.lua [hunk 10]
```diff
@@ -128,6 +128,140 @@ local function add_to_trust(playerName)
...
 end
 
+-- == RESEARCH PERMISSIONS ====================================================
+-- Uses a 'no_research' permission group (clone of Default minus research
+-- actions) to block tech-tree interactions for non-trusted players.
+-- Note: if the blueprint toggle in config.lua fires while research is locked,
+-- the no_research group may desync from Default for blueprint permissions.
+-- This is intentionally not handled — extremely unlikely during captain games.
+
+local RESEARCH_GROUP = 'no_research'
```
[the remark this answers is not in the document]
Comment (Darkenkade) [reply comfy_panel/special_games/captain.lua]: Fied

File: comfy_panel/special_games/captain.lua (hunk 4 above)
Comment (Darkenkade) [reply comfy_panel/special_games/captain.lua]: I have only copied what I saw all over bb base. Fixed.

File: comfy_panel/special_games/captain.lua [hunk 11]
```diff
@@ -1890,6 +2039,98 @@ local function on_gui_click(event)
...
+            get_dropdown_value(frame.captain_manager_research_trustlist_table.captain_add_research_trustlist_playerlist)
+        if playerNameUpdateText and playerNameUpdateText ~= '' and playerNameUpdateText ~= 'Select Player' then
+            local trustlist = special.northResearchTrustlist
+            local forceForPrint = 'north'
+            if player.name == special.captainList[2] then
+                trustlist = special.southResearchTrustlist
+                forceForPrint = 'south'
+            elseif player.name ~= special.captainList[1] then
+                player.print('You are not a captain', { color = Color.red })
+                return
```
[the remark this answers is not in the document]
Comment (Darkenkade) [reply comfy_panel/special_games/captain.lua]: Vice captains can access, there's a note for it.

File: comfy_panel/special_games/captain.lua (hunk 7 above)
Comment (Darkenkade) [reply comfy_panel/special_games/captain.lua]: This has been addressed properly in recent commits.

File: utils/datastore/jail_data.lua [hunk 12]
```diff
@@ -201,7 +201,7 @@ local teleport_player_to_gulag = function(player, action)
     elseif action == 'free' then
         local surface = game.surfaces[p_data.fallback_surface_index]
         local p = p_data.position
-        local p_group = game.permissions.get_group(p_data.p_group_id)
+        local p_group = game.permissions.get_group('Default')
```
[the remark this answers is not in the document]
Comment (Darkenkade) [reply utils/datastore/jail_data.lua]: It's handled correctly now by restoring saved group.

File: utils/datastore/jail_data.lua
[the remark this answers is not in the document]
Comment (Darkenkade) [reply utils/datastore/jail_data.lua:192]: We are back to using this so it's fine. It's used to restore spectators to spectator group on free.

File: comfy_panel/special_games/captain.lua (hunk 9 above)
Comment (Darkenkade) [reply comfy_panel/special_games/captain.lua]: Shortened message.

File: maps/biter_battles_v2/gui.lua [hunk 13]
```diff
@@ -1156,6 +1156,7 @@ function join_team(player, force_name, forced_join, auto_join)
     player.show_on_map = true
     Public.burners_balance(player)
     MultiSilo.on_player_changed_force(player)
+    Captain_event.on_player_joined_team(player)
```
[the remark this answers is not in the document]
Comment (Darkenkade) [reply maps/biter_battles_v2/gui.lua:1179]: Fixed

File: comfy_panel/special_games/captain.lua [hunk 14]
```diff
@@ -3851,10 +3797,24 @@ end
 
 function Public.reset_special_games()
     if storage.active_special_games.captain_mode then
+        cleanup_research_permissions()
```
[the remark this answers is not in the document]
Comment (Darkenkade) [reply comfy_panel/special_games/captain.lua]: Fixed now runs unconditionally.

**Comment by DaTa-** [comment]
@developer-8 PR looks good to me

It has some changes outside of it's own scope for the sake of being integrated into the scenario. It has some 1 issue due to a weird API behavior - research restricted group sometimes refuses to be deleted, despite no people in the group, so it just lingers harmlessly in permissions list. It also needs some more testing

**Review by developer-8 — APPROVED** [review]
I tested and it work
Please fix conflicts @Darkenkade

**Comment by developer-8** [comment]
Big PR it would be helpful if you also looked @cogito123

**Review by cogito123 — APPROVED** [review]
I looked at the code and tested a bit locally and not found any issue.  I have some concerns regarding how the code is structured, but the captain module in general is dumpster fire and this review is on going for almost two months so it's not reasonable to now request big changes.
